### ISSUE TITLE
For consideration: command overloads

### ISSUE THREAD

**Issue by Emzi0767** [NaamloosDT/DSharpPlus#172]
A feature requested semi-frequently is command overloads. This would allow having a single command that can take different combinations of arguments. With this, the following would be required:

* Instead of holding all properties directly, the `Command` entity would likely need to hold overload entities.
* Default help and related interfaces would need to be reworked to accommodate these changes.
* A priority attribute for overloads, which would define the order in which they'd be matched an executed.

Several CNext APIs could also use a facelift.

Edit: minor added note, support for `Nullable<T>`.

**Comment by NaamloosDT** [comment]
Though how do you want to do that? How do you want to check the difference between i.e. a string that represents an integer and a string that represents a ulong?

**Comment by Emzi0767** [comment]
I guess try binding arguments to the highest priority overload. If that fails, try next, rinse and repeat. If all fail, throw.

**Comment by NaamloosDT** [comment]
:thonkang:

**Comment by GHXX** [comment]
Sounds super useful to me! 😄

**Comment by audigex** [comment]
If nothing else, it would be great to be able to bind additional parameters

Eg I would like to have

`!vote no` and `!vote no 'somereason'` to allow people to add an additional parameter, but accept the vote without it

**Comment by uwx** [comment]
@audigex You can already do this just by using C# optional parameters.

**Comment by audigex** [comment]
Does that work? My code doesn't appear to work - complaining that "no" already exists

        [Command("no")]
        public async Task No(CommandContext ctx, int number)
        {
              //Something     
        }

        [Command("no")]
        public async Task No(CommandContext ctx, int number, string reason)
        {
              //Something else
        }

**Comment by uwx** [comment]
I mean something like this
```cs
    [Command("no")]
    public async Task No(CommandContext ctx, int number, string reason = "default value")
    {
          //Something
    }
```
But if you need significantly different functionality for both the methods then yes, command overloads are exactly the solution.

**Comment by audigex** [comment]
Ah okay - yeah that's handy for my specific use case :) thanks - although I'd still like to see it available if possible

**Comment by Emzi0767** [comment]
Implemented as of 5afb0800374df1da65fc6ae43f0aae291f68a12a.
